### DISCUSSION TITLE
[FR] A `dirty` property (or method) containing the pre-saved element

### DISCUSSION THREAD

**Discussion by lindseydiloreto** [craftcms/cms#8990] [Ideas]
When you save an element, there is a concept of "dirty" fields and attributes. This allows you to determine whether anything has changed during that particular save attempt.

```php
Event::on(
    Entry::class,
    Entry::EVENT_AFTER_SAVE,
    function (ModelEvent $e) {
        /* @var Entry $entry */
        $entry = $e->sender;

        // $entry->getDirtyFields();
        // $entry->getDirtyAttributes();
        // $entry->isFieldDirty();
        // $entry->isAttributeDirty();

    }
);
```

However, once you have determined that a particular field is dirty, _you still need to re-fetch the original data_ if you want to see what actually changed. In order to bridge this gap, I propose the following...

**A `dirty` property (or method) which contains the original pre-saved element.**

```php
Event::on(
    Entry::class,
    Entry::EVENT_AFTER_SAVE,
    function (ModelEvent $e) {
        /* @var Entry $entry */
        $entry = $e->sender;

        // Get the original entry
        $original = $entry->dirty();

    }
);
```

For reference, I discussed this with a few other smart folks [on Discord](https://discord.com/channels/456442477667418113/456442884258922529/801597572677632021). I'm probably understating the complexity of what's required to make it work, but it would be super sweet if this were physically possible!

**Comment by brandonkelly** [comment]
This is sortof the prerequisite to #6365.
